cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -48,7 +48,136 @@ if (MSVC)
+  elseif ("${COMPILER_FAMILY}" STREQUAL "gcc")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wall -Wconversion -Wno-sign-conversion")
+    # Treat all compiler warnings as errors
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wunknown-warning-option -Werror")
+  else()
+    message(FATAL_ERROR "Unknown compiler. Version info:\n${COMPILER_VERSION_FULL}")
+  endif()
+elseif ("${UPPERCASE_BUILD_WARNING_LEVEL}" STREQUAL "EVERYTHING")
+  # Pedantic builds for fixing warnings
+  if ("${COMPILER_FAMILY}" STREQUAL "msvc")
+    string(REPLACE "/W3" "" CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS}")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} /Wall")
+    # https://docs.microsoft.com/en-us/cpp/build/reference/compiler-option-warning-level
+    # /wdnnnn disables a warning where "nnnn" is a warning number
+    # Treat all compiler warnings as errors
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS}  /WX")
+  elseif ("${COMPILER_FAMILY}" STREQUAL "clang")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Weverything -Wno-c++98-compat -Wno-c++98-compat-pedantic")
+    # Treat all compiler warnings as errors
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Werror")
+  elseif ("${COMPILER_FAMILY}" STREQUAL "gcc")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wall -Wpedantic -Wextra -Wno-unused-parameter")
+    # Treat all compiler warnings as errors
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Werror")
+  else()
+    message(FATAL_ERROR "Unknown compiler. Version info:\n${COMPILER_VERSION_FULL}")
+  endif()
+else()
+  # Production builds (warning are not treated as errors)
+  if ("${COMPILER_FAMILY}" STREQUAL "msvc")
+    # https://docs.microsoft.com/en-us/cpp/build/reference/compiler-option-warning-level
+    # TODO: Enable /Wall and disable individual warnings until build compiles without errors
+    # /wdnnnn disables a warning where "nnnn" is a warning number
+    string(REPLACE "/W3" "" CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS}")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} /W3")
+  elseif ("${COMPILER_FAMILY}" STREQUAL "clang")
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Weverything -Wno-c++98-compat \
+    -Wno-c++98-compat-pedantic -Wno-deprecated -Wno-weak-vtables -Wno-padded \
+    -Wno-unused-parameter -Wno-undef -Wno-documentation-deprecated-sync \
+    -Wno-shadow -Wno-switch-enum -Wno-documentation -Wno-exit-time-destructors \
+    -Wno-global-constructors -Wno-weak-template-vtables -Wno-undefined-reinterpret-cast \
+    -Wno-implicit-fallthrough -Wno-old-style-cast -Wno-unreachable-code-return \
+    -Wno-float-equal -Wno-missing-prototypes -Wno-non-virtual-dtor \
+    -Wno-unused-macros -Wno-covered-switch-default -Wno-unreachable-code-break \
+    -Wno-extra-semi -Wno-cast-align -Wno-shift-sign-overflow \
+    -Wno-used-but-marked-unused -Wno-missing-variable-declarations \
+    -Wno-gnu-zero-variadic-macro-arguments -Wconversion -Wno-sign-conversion \
+    -Wno-disabled-macro-expansion -Wc++11-narrowing -Wnarrowing -Wno-shorten-64-to-32")
```

Comment:
If we are not passing `-Werror`, do we need all these suppressions in non-dev / checkin builds? A lot of these are pedantic warnings so it might be better to trim this down to a smaller list

Reply:
(i.e. not using `-Weverything` in prod / non-dev builds)

Reply:
I'll change that path back to "-Wall".

Reply:
I changed the production path back to "-Wall". Travis will add warnings to "-Wall" again.

Some warning suppressions subtracting from "-Weverything" were added for other versions of Clang used by Travis etc.. If Arrow developers are using newer versions of Clang, some of the warning suppressions can be removed; others require code changes.

BUILD_WARNING_LEVEL Values:

1) Checkin (use before checkin to attempt to determine if Travis CI build will fail)

This warning set subtracts from -Weverything by suppressing specific warning flags that prevent the current source code from building with various versions of Clang and adds warnings used by Travis, etc.

These are warnings that have never been enabled (Clang adds new warnings to -Weverthing ; -Wall is the list in common with gcc).

cmake -DBUILD_WARNING_LEVEL=checkin ..

2) Everything (use with the Clang compiler's -Weverything flag to find potential errors)

Use with BUILD_WARNING_FLAGS to disable warnings to get to the next warning.

http://amattn.com/p/better_apps_clang_weverything_or_wall_is_a_lie.html

cmake -DBUILD_WARNING_LEVEL=everything ..
make VERBOSE=1 unittest
cmake -DBUILD_WARNING_LEVEL=everything -DBUILD_WARNING_FLAGS="-Wno-padded -Wno-sign-conversion" ..
make VERBOSE=1 unittest
-DBUILD_WARNING_LEVEL=everything -DBUILD_WARNING_FLAGS="-Wno-padded -Wno-sign-conversion -Wno-unused-parameter" ..
make VERBOSE=1 unittest
etc.

3) Production (default; warnings are not treated as errors)

This warning set uses "-Wall". This is the default warning set to which Travis adds extra warnings and used by anyone compiling Arrow without specifying -DBUILD_WARNING_LEVEL when running cmake.

cmake ..